### PR TITLE
Delete Member and Host Feature

### PR DESCRIPTION
**Pull request by Kajol-Kumari**
Features: 
1. Remove member from participant team.
2. Remove host from Host team.

GIF of Delete Participant and Host Feature:
ezgif com-video-to-gif (18)

Requirement:
Need the backend changes done in PR https://github.com/Cloud-CV/EvalAI/pull/2660 for functionality to work correctly.

### CODE REVIEW

**Comment by Kajol-Kumari** [comment]
@Suryansh5545 please take a look.

File: src/app/services/endpoints.service.ts [hunk 1]
```diff
@@ -97,6 +97,13 @@ export class EndpointsService {
     return `${this.participants}participant_team/${teamId}`;
   }
 
+  /**
+   * Delete Member Participant
+   */
+  participantMemberURL(teamId) {
+    return `${this.participants}remove_member_from_participant_team/${teamId}`;
```
Comment (Suryansh5545) [inline src/app/services/endpoints.service.ts]: @kajol-kumari I checked the participants url file and there is no "remove_member_from_participant_team" Because of which you are recieving not found error

Comment (Kajol-Kumari) [reply src/app/services/endpoints.service.ts]: > @Kajol-Kumari I checked the participants url file and there is no "remove_member_from_participant_team" Because of which you are recieving not found error

@Suryansh5545  can you tell me which url file you are taking about as i couldn't find the "remove_member_from_participant_team" either except teamlist.component.ts

Comment (Suryansh5545) [reply src/app/services/endpoints.service.ts]: @kajol-kumari  https://github.com/Cloud-CV/EvalAI/blob/master/apps/participants/urls.py

Comment (Suryansh5545) [reply src/app/services/endpoints.service.ts]: Did you get it?

Comment (Kajol-Kumari) [reply src/app/services/endpoints.service.ts]: yes @Suryansh5545 got it. Thanks!

**Comment by Kajol-Kumari** [comment]
@Sanji515 @Suryansh5545 Please take a look.

**Comment by Suryansh5545** [comment]
@Sanji515 Sir, I won't be able to complete this task #276 because of my exams before 6 march . So I would suggest  if @Kajol-Kumari agrees then she takes over it . As she is already working on teamlist folder.

**Comment by Kajol-Kumari** [comment]
> @Sanji515 Sir, I won't be able to complete this task #276 because of my exams before 6 march . So I would suggest if @Kajol-Kumari agrees then she takes over it . As she is already working on teamlist folder.

If @Sanji515 sir permits , i can continue to work on #276 . Also sir, please let us know whether you want the work in single PR or two.

**Comment by Sanji515** [comment]
Sure @Kajol-Kumari go ahead. As it is almost done so please try to build on top of that PR and open a new PR with the joint commit.

**Comment by Kajol-Kumari** [comment]
> Sure @Kajol-Kumari go ahead. As it is almost done so please try to build on top of that PR and open a new PR with the joint commit.

Okay @Sanji515 sir.

**Comment by Kajol-Kumari** [comment]
@Sanji515 I am closing this PR and opening a new PR having feature of both #288 and #276

**Comment by Sanji515** [comment]
@Kajol-Kumari If you have already built these two features in the same branch then it's fine to open a single PR but if you haven't started yet then it would be better to open separate PR for these two.

**Comment by Kajol-Kumari** [comment]
> @Kajol-Kumari If you have already built these two features in the same branch then it's fine to open a single PR but if you haven't started yet then it would be better to open separate PR for these two.

@Sanji515 Sir, I have already merged the changes in a single branch.

**Comment by Sanji515** [comment]
No problem, please open a PR containing both features.

**Comment by Kajol-Kumari** [comment]
@Sanji515 Sir, I have separated both the PRs as it had become huge after combining them.

**Comment by Kajol-Kumari** [comment]
@Sanji515  @Shekharrajak @RishabhJain2018 can you please review this PR?

**Comment by Shekharrajak** [comment]
@Sanji515  can review this PR better.

File: src/app/components/publiclists/teamlist/teamcard/teamcard.component.html [hunk 2]
```diff
@@ -1,41 +1,44 @@
...
+        </span>
+        <br>
+        <span class="title">
+        Created by:
+        </span>
+        <span>
+        {{teamView['created_by']}}
+        </span>
+      </div>
+      <div class='team-icons'>
```
Comment (Sanji515) [inline src/app/components/publiclists/teamlist/teamcard/teamcard.component.html]: I don't think there is any change to these lines, can you please revert the changes here @Kajol-Kumari

File: src/app/components/publiclists/teamlist/teamcard/teamcard.component.ts [hunk 3]
```diff
@@ -13,6 +14,21 @@ import { Router, ActivatedRoute } from '@angular/router';
 })
 export class TeamcardComponent implements OnInit, OnChanges {
 
+  /**
+   * User Name
+   */
+  user = {username: ''};
```
Comment (Sanji515) [inline src/app/components/publiclists/teamlist/teamcard/teamcard.component.ts]: Please change the variable to `LoggedInUser` or something more sensible and also change the comment to `LoggedIn User Details`

File: src/app/components/publiclists/teamlist/teamcard/teamcard.component.ts [hunk 4]
```diff
@@ -13,6 +14,21 @@ import { Router, ActivatedRoute } from '@angular/router';
...
+
+  /**
+   * Authentication Service subscription
+   */
+  authServiceSubscription: any;
+
+  /**
+   * Current Authentication state
+   */
+  authState: any;
```
Comment (Sanji515) [inline src/app/components/publiclists/teamlist/teamcard/teamcard.component.ts:25]: Do we need this variable? or you can keep this by removing [this](https://github.com/Cloud-CV/EvalAI-ngx/pull/288/files?file-filters%5B%5D=.html&file-filters%5B%5D=.ts&file-filters%5B%5D=.yml#diff-39040dd12941e102fae3dbad153f9262R20) one

File: src/app/components/publiclists/teamlist/teamcard/teamcard.component.ts [hunk 5]
```diff
@@ -132,6 +170,15 @@ export class TeamcardComponent implements OnInit, OnChanges {
     this.deleteTeamCard.emit(this.team['id']);
   }
 
+    /**
+   * Fires delete member event.
+   */
+  deleteTeamMember(e, participantId) {
+    e.stopPropagation();
+    const teamid = this.team['id'];
```
Comment (Sanji515) [inline src/app/components/publiclists/teamlist/teamcard/teamcard.component.ts]: Can we remove this extra line and pass `this.team['id]` directly to the emit function as we are not using it anywhere else

File: src/app/components/publiclists/teamlist/teamcard/teamcard.component.ts [hunk 6]
```diff
@@ -156,17 +203,26 @@ export class TeamcardComponent implements OnInit, OnChanges {
...
     if (memberString !== '') {
       memberString = memberString.slice(2, memberString.length);
     }
+    if (memberIdString !== '') {
+      memberIdString = memberIdString.slice(2, memberIdString.length);
+    }
     this.teamView['members'] = memberString;
+    this.teamView['member_ids'] = memberIdString;
+    this.memberArray = memberString.split(',');
+    this.memberIdArray = memberIdString.split(', ');
```
Comment (Sanji515) [inline src/app/components/publiclists/teamlist/teamcard/teamcard.component.ts]: Can we change the logic of doing like this? I think at the end we need an array with member and memberId array so how about directly initializing an array and push the value to it rather than initializing an string and then appening `,`(comma) and then value @Kajol-Kumari What do you think?

File: src/app/components/publiclists/teamlist/teamlist.component.ts [hunk 7]
```diff
@@ -227,10 +237,12 @@ export class TeamlistComponent implements OnInit, OnDestroy {
       this.fetchTeamsPath = 'hosts/challenge_host_team';
       this.createTeamsPath = 'hosts/create_challenge_host_team';
       this.deleteTeamsPath = 'hosts/remove_self_from_challenge_host';
+      this.deleteMembersPath = 'hosts/challenge_host_team/';
       this.fetchMyTeams(this.fetchTeamsPath);
       this.teamCreateTitle = 'Create a New Team';
       this.teamSelectTitle = 'Select a Challenge Host Team';
       this.teamCreateButton = 'Create Host Team';
+      this.MemberDeleteTitle = '/challenge_host/';
```
Comment (Sanji515) [inline src/app/components/publiclists/teamlist/teamlist.component.ts]: I think this is not an title so how about if we declare only one variable for path to something like this 'hosts/challenge_host_team/<team_id>/challenge_host/' and then [here](https://github.com/Cloud-CV/EvalAI-ngx/pull/288/files?file-filters%5B%5D=.html&file-filters%5B%5D=.ts&file-filters%5B%5D=.yml#diff-cb564c2cd5f0dfd1606b8de1c85f077aR590) we can just replace this `<team_id>` with the team ID. @Kajol-Kumari

File: src/app/components/publiclists/teamlist/teamcard/teamcard.component.ts (hunk 6 above)
Comment (Kajol-Kumari) [reply src/app/components/publiclists/teamlist/teamcard/teamcard.component.ts]: > Can we change the logic of doing like this? I think at the end we need an array with member and memberId array so how about directly initializing an array and push the value to it rather than initializing an string and then appening `,`(comma) and then value @Kajol-Kumari What do you think?

@Sanji515 I am doing it this way as i need to store the members in the form of `,` separated string in the `teamView` as it is being used to display the list of member. So, i need both comma separated string as well as array and so i am doing it this way.

Comment (Sanji515) [reply src/app/components/publiclists/teamlist/teamcard/teamcard.component.ts]: But now I think you are displaying as a list with two columns one with the member username and another with delete icon. I don't see any comma separated text in UI.

File: src/app/components/publiclists/teamlist/teamlist.component.ts (hunk 7 above)
Comment (Kajol-Kumari) [reply src/app/components/publiclists/teamlist/teamlist.component.ts]: @Sanji515  as [here](https://github.com/Cloud-CV/EvalAI-ngx/pull/288/files#diff-cb564c2cd5f0dfd1606b8de1c85f077aR259) you can see i am resetting the value of  `MemberDeleteTitle` as this url is being used for both host and participant member deletion so i don't think we can't fix the path like this(^).

Comment (Sanji515) [reply src/app/components/publiclists/teamlist/teamlist.component.ts]: Yes but still I think we can do this way, that single variable will be updated according to the conditions which you have pointed.

File: src/app/components/publiclists/teamlist/teamcard/teamcard.component.ts [hunk 8]
```diff
@@ -155,18 +196,19 @@ export class TeamcardComponent implements OnInit, OnChanges {
...
+        this.memberArray.push(temp[i]['user']);
+        this.memberIdArray.push(temp[i]['id']);
       }
     }
-    if (memberString !== '') {
-      memberString = memberString.slice(2, memberString.length);
-    }
-    this.teamView['members'] = memberString;
+    this.teamView['members'] = this.memberArray;
+    this.teamView['member_ids'] = this.memberIdArray;
```
Comment (Sanji515) [inline src/app/components/publiclists/teamlist/teamcard/teamcard.component.ts:208]: Yeah now it looks clean :smiley:

File: src/app/components/publiclists/teamlist/teamcard/teamcard.component.html (hunk 2 above)
Comment (Sanji515) [reply src/app/components/publiclists/teamlist/teamcard/teamcard.component.html]: I think you missed this @Kajol-Kumari

File: src/app/components/publiclists/teamlist/teamcard/teamcard.component.ts [hunk 9]
```diff
@@ -85,6 +112,12 @@ export class TeamcardComponent implements OnInit, OnChanges {
    */
   ngOnInit() {
     this.updateView();
+    this.authServiceSubscription = this.authService.change.subscribe((authState) => {
+      this.authState = authState;
+      if (this.authState.isLoggedIn) {
+        this.authState = authState;
```
Comment (Sanji515) [inline src/app/components/publiclists/teamlist/teamcard/teamcard.component.ts]: I didn't get this thing, the way it it written currently I don't find any use of `if` condition here cause in line 116 it is already assigning to the variable. @Kajol-Kumari

File: src/app/components/publiclists/teamlist/teamcard/teamcard.component.html (hunk 2 above)
Comment (Kajol-Kumari) [reply src/app/components/publiclists/teamlist/teamcard/teamcard.component.html]: @Sanji515 I tried reverting back the changes here and also i have't done any changes in these lines but still it's appearing like this.Don't know why.

File: src/app/components/publiclists/teamlist/teamcard/teamcard.component.ts (hunk 9 above)
Comment (Kajol-Kumari) [reply src/app/components/publiclists/teamlist/teamcard/teamcard.component.ts]: yes @Sanji515 I had applied this check for getting username but now i have replaced it with `authstate` variable so yeah we don't need this condition check anymore.

File: src/app/components/publiclists/teamlist/teamcard/teamcard.component.html [hunk 10]
```diff
@@ -1,41 +1,48 @@
...
-	    {{teamView['team_name']}}
-      </span>
-      <br>
-      <span class="title">
-	    Created by:
-      </span>
-      <span>
-	    {{teamView['created_by']}}
-      </span>
+<div class="team-card-container" [class.selected]='isSelected' [class.noexpand]='isHost' [class.expand]='!isHost'>
```
Comment (Sanji515) [inline src/app/components/publiclists/teamlist/teamcard/teamcard.component.html]: @Kajol-Kumari Here you can checkout this file as in the master and then do your changes only.

File: src/app/components/publiclists/teamlist/teamlist.component.ts [hunk 11]
```diff
@@ -565,4 +572,39 @@ export class TeamlistComponent implements OnInit, OnDestroy {
...
+        },
+        err => {
+          SELF.globalService.handleApiError(err);
+        },
+        () => {}
+        );
+      };
+      const PARAMS = {
+        title: 'Would you like to remove this member ?',
+        content: 'Note: This action will remove this member from the team.',
```
Comment (Sanji515) [inline src/app/components/publiclists/teamlist/teamlist.component.ts:599]: @Kajol-Kumari Don't you think we should also modify this message when host tries to remove himself as it will delete the team directly instead of removing the host only.

**Review by Sanji515 — COMMENTED** [review]
Minor comments, Rest LGTM :+1:

File: src/app/components/publiclists/teamlist/teamlist.component.ts (hunk 11 above)
Comment (Kajol-Kumari) [reply src/app/components/publiclists/teamlist/teamlist.component.ts:599]: @Sanji515 If a host tries to remove himself from the team. The request will be aborted and a message is being shown saying that `if you want to delete yourself, instead you delete the team` as shown in the GIF above.

Comment (Sanji515) [reply src/app/components/publiclists/teamlist/teamlist.component.ts:599]: Yeah but I've just checked on my local instead of showing the message it directly deletes the team saying `Member removed from the team!`. Can you please check this once @Kajol-Kumari

Comment (Kajol-Kumari) [reply src/app/components/publiclists/teamlist/teamlist.component.ts:599]: yes @Sanji515 that is because you are using the current EvalAI backend and it has a bug that i have mentioned in [here](https://github.com/Cloud-CV/EvalAI/issues/2664). I have fixed this in [this](https://github.com/Cloud-CV/EvalAI/pull/2781) PR and i have used that branch as backend for working on this PR. I have also mentioned it in the PR description. Please take a look at it.

Comment (Sanji515) [reply src/app/components/publiclists/teamlist/teamlist.component.ts:599]: I see. But can't we add that check on the frontend itself. As we know who the host is and then we can just check the user if it's matches to the host user.

Comment (Sanji515) [reply src/app/components/publiclists/teamlist/teamlist.component.ts:599]: okay lets add this in a separate PR as it may require some other variables for the check.

**Review by Sanji515 — APPROVED** [review]
LGTM! :+1: This can be merged now :tada:
